pkg/agent/flowexporter/exporter/exporter.go

```diff
@@ -147,57 +146,153 @@ func NewFlowExporter(connStore *connections.ConntrackConnectionStore, records *f
+			nextExpireTime, nextIsDeny, err = exp.sendFlowRecords(nextIsDeny)
+			if err != nil {
+				klog.Errorf("Error when sending expired flow records: %v", err)
+				// If there is an error when sending flow records because of intermittent
+				// connectivity, we reset the connection to IPFIX collector and retry
+				// in the next export cycle to reinitialize the connection and send flow records.
+				exp.process.CloseConnToCollector()
+				exp.process = nil
+				expireTimer.Reset(defaultTimeout)
+				continue
+			}
+			expireTimer.Reset(nextExpireTime)
 		}
 	}
-	// Send flow records to IPFIX collector.
-	err := exp.sendFlowRecords()
-	if err != nil {
-		klog.Errorf("Error when sending flow records: %v", err)
-		// If there is an error when sending flow records because of intermittent connectivity, we reset the connection
-		// to IPFIX collector and retry in the next export cycle to reinitialize the connection and send flow records.
-		exp.process.CloseConnToCollector()
-		exp.process = nil
-		return
-	}
-	klog.V(2).Infof("Successfully exported IPFIX flow records")
+}
+
+func (exp *flowExporter) sendFlowRecords(nextIsDeny bool) (time.Duration, bool, error) {
+	expiredConns := make([]flowexporter.Connection, 0)
+	currTime := time.Now()
+	var conntrackExpireTime, denyExpireTime time.Duration
+
+	getTopExpiredItem := func(pq *priorityqueue.ExpirePriorityQueue) *flowexporter.ItemToExpire {
+		// If the queue is empty or top item is not timeout, then we do not have to
+		// check the following items.
+		topItem := pq.Peek()
+		if topItem == nil || (topItem.ActiveExpireTime.After(currTime) && topItem.IdleExpireTime.After(currTime)) {
+			return nil
+		}
+		pqItem := heap.Pop(pq).(*flowexporter.ItemToExpire)
+		expiredConns = append(expiredConns, *pqItem.Conn)
+		return pqItem
+	}
+	if !nextIsDeny {
+		exp.conntrackConnStore.AcquireConnStoreLock()
+		for {
+			pqItem := getTopExpiredItem(exp.conntrackPriorityQueue)
+			if pqItem == nil {
+				break
+			}
+			if flowexporter.IsConnectionDying(pqItem.Conn) {
+				// If a conntrack connection is in dying state or connection
+				// is not in conntrack table, we set the ReadyToDelete flag to
+				// true to do the deletion later.
+				pqItem.Conn.ReadyToDelete = true
+			}
+			if pqItem.IdleExpireTime.Before(currTime) {
+				// If a conntrack connection item is idle time out, we consider
+				// the connection to be inactive, as no related flows coming within
+				// the idle time out range.
+				pqItem.Conn.IsActive = false
+			}
+			exp.conntrackConnStore.UpdateConnStoreAndQueue(pqItem, currTime)
+		}
+		conntrackExpireTime = exp.conntrackPriorityQueue.GetExpiryFromExpirePriorityQueue()
+		exp.conntrackConnStore.ReleaseConnStoreLock()
+		exp.denyConnStore.AcquireConnStoreLock()
+		denyExpireTime = exp.denyPriorityQueue.GetExpiryFromExpirePriorityQueue()
+		exp.denyConnStore.ReleaseConnStoreLock()
+	} else {
+		exp.denyConnStore.AcquireConnStoreLock()
```

Comment:
I can't figure out a way to let conntrack connection store and deny connection store share a same `getExpiredConns`, unless we make `connectionStore` as a public field of conntrack/deny connection store. Please checkout the latest commit to see whether that looks good. Thanks!

Reply:
I think you can make `connectionStore` public which was my first suggestion

My second suggestion ("introducing an interface") may better:
```go
type ConnectionStoreInterface interface {
      sync.Locker
      UpdateStoreAndQueue(...)
      // any other method that may improve code reusability
}
```

Reply:
Thanks Antonin for the explanation. I added an interface. Two things I'm not sure about - 

- Since part of the logic of the `getExpiredConns` is different between conntrack connection and deny connection(the criterias to determine whether a conn is active/ready to delete), I implemented the method separately in conntrack_connections.go and deny_connections.go. The lines of code actually do not decrease in this case. Is that expected?
- If we have to implement the methods separately in two files, what's the major benefit of having an interface compared to without an interface, in our case?